### PR TITLE
[TACHYON-274] WebUI Improvement about Hierarchy

### PR DESCRIPTION
**Pull request by Yours-shupeng**
qq 20150209132408
We improve the overview page of webUI about Hierarchy to add information about "Hierarchy Store Level" "Hierarchy Store Alias" "Space Capacity" "Space Used" and "Space Usage" information. In order to realize the function, we amend tachyon.thrift file to recreate tachyon.thrift package. Maily,  we amend worker_register and worker_heartbeat function to replace total bytes and used bytes of workers with hierarchy store levels, hierarchy store alias values, total bytes of each hierarchy store level, used bytes of  each hierarchy store level. In MasterInfo class, we add getHierarchyStoreLevels(), getHierarchyStoreLevelAliasValues(), getHierarchyCapacityBytes(), getHierarchyUsedBytes() to collect hierarchy information about workers.

### CODE REVIEW

File: core/src/thrift/tachyon.thrift [hunk 1]
```diff
@@ -138,8 +138,9 @@ service MasterService {
    * is master started time. currentBlocks maps from id of storage directory to the blocks it
    * contains.
    */
-  i64 worker_register(1: NetAddress workerNetAddress, 2: i64 totalBytes, 3: i64 usedBytes,
-      4: map<i64, list<i64>> currentBlocks)
+  i64 worker_register(1: NetAddress workerNetAddress, 2: list<i32> storageLevels,
+      3: list<i32> storageLevelAliasValues, 4: list<i64> totalBytes, 5: list<i64> usedBytes,
+      6: map<i64, list<i64>> currentBlocks)
```
Comment (haoyuan) [inline core/src/thrift/tachyon.thrift]: why make this change?

@RongGu

**Comment by haoyuan** [comment]
@shimingfei @GraceH

**Comment by Yours-shupeng** [comment]
We can't collect hierarchy information in MasterInfo class now, so we want to workers send hierarchy information to master and update it. The work_register, worker_heartbeat(), cache_Block() methods update workers information, so we want these classes update hierarchy information at the same time and update the web information about hierarchy.

**Comment by dcapwell** [comment]
"We can't collect hierarchy information in MasterInfo class now, so we want to workers send hierarchy information to master and update it"

Why put it in master? This
Increases the amount of memory the master needs and since each node can have a completely different setup (not all nodes have identical hardware or things running on it) the ui will get harder to understand.

What do we gain by putting this information on the master?

**Comment by Yours-shupeng** [comment]
I think this information is meaningful because it makes users know the usage information of mem, ssd and hdd in detail. While we have to collect hierarchy information about workers in master so as to show it in web pages. We think all workers use default same configuration in this PR and didn't consider different setup of workers, so we would adjust the hierarchy store level information to information about MEM, SSD and HDD.  We will detail the hierarchy information of each worker in next PR. Thanks for you review.

**Comment by dcapwell** [comment]
" I think this information is meaningful because it makes users know the
usage information of mem, ssd and hdd in detail"

I fully agree.  Monitoring systems are a great place for this information
and do a really good job giving incite into how the storage is getting
used.  We can solve this problem with adding monitoring and use existing
tools rather than start to build the same feature sets in the master.

Do you have other use cases in mind for why this data is needed in the
master?  I don't really buy the idea of adding it for display purposes only.

**Comment by shimingfei** [comment]
@dcapwell Do you mean the information can be collected by an existing independent module such as metric system instead of adding information into master?

**Comment by dcapwell** [comment]
Yes.  If you push this information into existing metrics systems you will
get the same feature set and then some.  This also puts less pressure on
the master

**Pull request closed without merging**